### PR TITLE
[New model] Add code for BEVFusion

### PR DESCRIPTION
**Pull request by Birdylx**

### CODE REVIEW

File: paddle3d/apis/trainer.py [hunk 1]
```diff
@@ -260,13 +260,15 @@ def train(self):
                     self.model)
 
         model = self.model
+        logger.info(model)
```
Comment (FlyingQianMM) [inline paddle3d/apis/trainer.py]: 是否有必要打印model?

File: paddle3d/apis/pipeline.py [hunk 2]
```diff
@@ -82,15 +107,16 @@ def training_step(model: paddle.nn.Layer,
...
+    # reduce loss when distributed training
+    if paddle.distributed.is_initialized():
+        with paddle.no_grad():
+            for loss_name, loss_value in log_loss.items():
+                loss_clone = loss_value.clone()
+                paddle.distributed.all_reduce(
+                    loss_clone.scale_(1. / paddle.distributed.get_world_size()))
+                log_loss[loss_name] = loss_clone.item()
+
+    return log_loss, grad_norm
```
Comment (FlyingQianMM) [inline paddle3d/apis/pipeline.py:112]: 多卡同步loss用于显示，以及计算梯度都会造成训练速度变慢，可考虑去掉

File: paddle3d/models/detection/bevfusion/bevf_transforms.py [hunk 3]
```diff
@@ -0,0 +1,1174 @@
+# Copyright (c) 2022 PaddlePaddle Authors. All Rights Reserved.
```
Comment (FlyingQianMM) [inline paddle3d/models/detection/bevfusion/bevf_transforms.py:1]: bevfusion的所有transforms都重新写了一套？看有些transforms不能复用之前的原因是啥

File: paddle3d/models/detection/bevfusion/loss.py [hunk 4]
```diff
@@ -0,0 +1,362 @@
...
+# limitations under the License.
+
+import paddle
+import paddle.nn as nn
+import paddle.nn.functional as F
+
+from paddle3d.apis import manager
+
+__all__ = ['FocalLoss', 'SmoothL1Loss', 'CrossEntropyLoss']
+
```
Comment (FlyingQianMM) [inline paddle3d/models/detection/bevfusion/loss.py]: Loss这部分也有重合

File: paddle3d/models/layers/param_init.py [hunk 5]
```diff
@@ -22,6 +22,12 @@
 from paddle3d.utils.logger import logger
 
 
+def bias_init_with_prob(prior_prob):
+    """initialize conv/fc bias value according to a given probability value."""
```
Comment (FlyingQianMM) [inline paddle3d/models/layers/param_init.py]: 该文件内已有init_bias_by_prob

File: paddle3d/models/voxel_encoders/voxel_encoder.py [hunk 6]
```diff
@@ -51,3 +55,229 @@ def forward(self, features, num_voxels, coors=None):
...
+        else:
+            # [K, 1, units]
+            repeated = aggregated.tile([1, voxel_count, 1])
+            concatenated = paddle.concat([pointwise, repeated], axis=2)
+            # [K, T, 2 * units]
+            return concatenated
+
+
+@manager.VOXEL_ENCODERS.add_component
+class HardVFE(nn.Layer):
```
Comment (FlyingQianMM) [inline paddle3d/models/voxel_encoders/voxel_encoder.py:142]: 这个看着也有点重复实现了

File: paddle3d/models/voxelizers/voxelize_v2.py [hunk 7]
```diff
@@ -0,0 +1,80 @@
...
+        if isinstance(max_num_voxels, (tuple, list)):
+            self.max_num_voxels = max_num_voxels
+        else:
+            self.max_num_voxels = [max_num_voxels, max_num_voxels]
+
+    def single_forward(self, point, max_num_voxels, bs_idx):
+        voxels, coors, num_points_per_voxel, voxels_num = voxelize_v2.hard_voxelize_v2(
+            point, self.voxel_size, self.point_cloud_range,
+            self.max_num_points_in_voxel, max_num_voxels)
+        paddle.device.cuda.synchronize()
```
Comment (FlyingQianMM) [inline paddle3d/models/voxelizers/voxelize_v2.py]: HardVoxelize换成Paddle3D的实现。这个版本很慢，不需要paddle.device.cuda.synchronize()

File: paddle3d/utils/sync_bn.py [hunk 8]
```diff
@@ -0,0 +1,114 @@
...
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+import paddle
+import paddle.distributed as dist
+import paddle.nn as nn
+from paddle.autograd import PyLayer
+
+__all__ = ['NaiveSyncBatchNorm1D', 'NaiveSyncBatchNorm2D']
```
Comment (FlyingQianMM) [inline paddle3d/utils/sync_bn.py]: 这两个类跟Paddle的syncbn有什么区别？

File: paddle3d/models/detection/bevfusion/utils.py [hunk 9]
```diff
@@ -0,0 +1,275 @@
...
+        cls_inds = paddle.nonzero(mlvl_scores[:, i] > score_thr)
+        if cls_inds.numel() == 0:
+            continue
+
+        _scores = paddle.gather(mlvl_scores[:, i], cls_inds)
+        _bboxes_for_nms = paddle.gather(mlvl_bboxes_for_nms, cls_inds)
+        order = _scores.argsort(0, descending=True)
+        _bboxes_for_nms = paddle.gather(_bboxes_for_nms, index=order)
+
+        keep, num_out = nms_gpu_v2(_bboxes_for_nms, cfg['nms_thr'])
```
Comment (FlyingQianMM) [inline paddle3d/models/detection/bevfusion/utils.py]: nms_gpu可复用已有的

File: paddle3d/apis/trainer.py (hunk 1 above)
Comment (Birdylx) [reply paddle3d/apis/trainer.py]: 可删除

File: paddle3d/apis/pipeline.py (hunk 2 above)
Comment (Birdylx) [reply paddle3d/apis/pipeline.py:112]: 用于显示正确的多卡loss值，这块时间消耗应该可以忽略吧。grad_norm不计算梯度，这块用于验证梯度裁剪，可删除。

File: paddle3d/models/detection/bevfusion/bevf_transforms.py (hunk 3 above)
Comment (Birdylx) [reply paddle3d/models/detection/bevfusion/bevf_transforms.py:1]: 是的，之前同步开发，部分transforms不能复用，其他的还需要验证。

File: paddle3d/models/detection/bevfusion/loss.py (hunk 4 above)
Comment (Birdylx) [reply paddle3d/models/detection/bevfusion/loss.py]: loss复用还需要验证。

File: paddle3d/models/layers/param_init.py (hunk 5 above)
Comment (Birdylx) [reply paddle3d/models/layers/param_init.py]: 已修改。

File: paddle3d/models/voxel_encoders/voxel_encoder.py (hunk 6 above)
Comment (Birdylx) [reply paddle3d/models/voxel_encoders/voxel_encoder.py:142]: 快速对齐

File: paddle3d/models/voxelizers/voxelize_v2.py (hunk 7 above)
Comment (Birdylx) [reply paddle3d/models/voxelizers/voxelize_v2.py]: 可复用，已删除冗余代码

File: paddle3d/utils/sync_bn.py (hunk 8 above)
Comment (Birdylx) [reply paddle3d/utils/sync_bn.py]: 用于对齐，后续未使用，已删除冗余代码。

File: paddle3d/models/detection/bevfusion/utils.py (hunk 9 above)
Comment (Birdylx) [reply paddle3d/models/detection/bevfusion/utils.py]: 验证中。

File: configs/bevfusion/cam_stream/bevf_pp_4x8_2x_nusc_cam.yaml [hunk 10]
```diff
@@ -0,0 +1,169 @@
...
+
+model:
+  type: BEVF_FasterRCNN
+  camera_stream: true
+  lss: false
+  grid: 0.5
+  num_views: 6
+  final_dim: [900, 1600]
+  downsample: 8
+  load_img_from: /workspace/liuxiao28/Paddle3D/mask_rcnn_dbswin-t_fpn_3x_nuim_cocopre.pdparams
```
Comment (LielinJiang) [inline configs/bevfusion/cam_stream/bevf_pp_4x8_2x_nusc_cam.yaml]: 使用相对路径吧

File: paddle3d/models/detection/bevfusion/bevf_transforms.py (hunk 3 above)
Comment (Birdylx) [reply paddle3d/models/detection/bevfusion/bevf_transforms.py:1]: 部分transform已修改和复用

File: paddle3d/models/detection/bevfusion/loss.py (hunk 4 above)
Comment (Birdylx) [reply paddle3d/models/detection/bevfusion/loss.py]: loss已复用

File: paddle3d/models/detection/bevfusion/utils.py (hunk 9 above)
Comment (Birdylx) [reply paddle3d/models/detection/bevfusion/utils.py]: nms_gpu已复用

File: configs/bevfusion/cam_stream/bevf_pp_4x8_2x_nusc_cam.yaml (hunk 10 above)
Comment (Birdylx) [reply configs/bevfusion/cam_stream/bevf_pp_4x8_2x_nusc_cam.yaml]: 已修改，这是一阶段训练需要的权重，由用户自行设置，在README里进行了说明。

File: configs/bevfusion/bevf_pp_2x8_1x_nusc.yaml [hunk 11]
```diff
@@ -0,0 +1,217 @@
...
+      std: [58.395, 57.12, 57.375]
+      to_rgb: true
+    - type: PadImage
+      size_divisor: 32
+    - type: SampleFilterByKey
+      keys: ['points', 'img']
+  mode: val
+
+model:
+  type: BEVF_FasterRCNN
```
Comment (LielinJiang) [inline configs/bevfusion/bevf_pp_2x8_1x_nusc.yaml]: 类名风格和已有的有点不同，建议是BEVFFasterRCNN，不带下划线

Comment (Birdylx) [reply configs/bevfusion/bevf_pp_2x8_1x_nusc.yaml]: Done
